client/src/controllers/TabsController.ts

```diff
@@ -0,0 +1,319 @@
+/* eslint-disable no-shadow */
+import { Controller } from '@hotwired/stimulus';
+import { debounce } from '../utils/debounce';
+
+interface IndexedEventTarget extends EventTarget {
+  index: number;
+}
+
+interface TabLink extends HTMLAnchorElement {
+  index: number;
+}
+
+/**
+ * @example - creating a simple tabs interface
+ *<div class="w-tabs" data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-selected-value="" data-w-tabs-selected-class="animate-in">
+ *    <div class="w-tabs__list" role="tablist" data-w-tabs-target="list">
+ *        <a id="tab-label-tab-1" href="#tab-tab-1" class="w-tabs__tab" role="tab" tabindex="-1"
+ *           data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast" data-w-tabs-target="label">
+ *            Tab 1
+ *        </a>
+ *        <a id="tab-label-tab-2" href="#tab-tab-2" class="w-tabs__tab" role="tab"
+ *           data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast" data-w-tabs-target="label">
+ *            Tab 2
+ *        </a>
+ *    </div>
+ *    <div class="tab-content tab-content--comments-enabled">
+ *        <section id="tab-tab-1" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
+ *            tab-1
+ *        </section>
+ *        <section id="tab-tab-2" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
+ *            tab-2
+ *        </section>
+ *    </div>
+ *</div>
+ */
+
+export class TabsController extends Controller<HTMLDivElement> {
+  static targets = ['list', 'trigger', 'label', 'panel'];
+
+  static classes = ['selected'];
+
+  static values = {
+    transition: { default: 150, type: Number },
+    selected: { default: '', type: String },
+    syncURLHash: { default: false, type: Boolean },
+    animate: { default: true, type: Boolean },
+  };
+
+  declare readonly selectedClasses: string;
+
+  declare listTarget: HTMLDivElement;
+  declare triggerTargets: HTMLAnchorElement[];
+  declare labelTargets: HTMLAnchorElement[];
+  declare panelTargets: HTMLElement[];
+
+  declare syncURLHashValue: boolean;
+  declare selectedValue: string;
+  declare transitionValue: number;
+  declare animateValue: boolean;
```

Comment:
Let's clean this up, using JSDoc to describe each value. Super helpful for contributors and documentation.

I still think we can streamline the terminology throughout (selected/active) and also how the animation works but let's leave the variables as is for now.

The only removal here is the list target.

Also, note that classes will be an array of strings - https://stimulus.hotwired.dev/reference/css-classes 

```suggestion
  /** ID of the currently selected tab. */
  declare selectedValue: string;
  /** If true, animation will run when a new tab is selected. */
  declare readonly animateValue: boolean;
  /** Tab elements, with role='tab', allowing switching between tabs. */
  declare readonly labelTargets: HTMLAnchorElement[];
  /** Tab content panels, with role='tabpanel', showing the content for each tab. */
  declare readonly panelTargets: HTMLElement[];
  /** Other elements within the controller's scope that may trigger a specific tab. */
  declare readonly triggerTargets: HTMLAnchorElement[];
  /** Classes to set on the tab panel content when selected. */
  declare readonly selectedClasses: string[];
  /** If true, the selected tab will sync with the URL hash. */
  declare readonly syncURLHashValue: boolean;
  /** The time in milliseconds for the tab content to transition in and out. */
  declare readonly transitionValue: number;
```